libc/stdio_file.c

```diff
@@ -0,0 +1,785 @@
+		TEST_ASSERT_EQUAL_INT('b', getc(fd));
+		TEST_ASSERT_EQUAL_INT(EOF, fgetc(fd));
+		TEST_ASSERT_EQUAL_INT(EOF, getc(fd));
+	}
+	fclose(fd);
+
+	/* read from file open for writing */
+	fd = fopen(stdpath, "w");
+	{
+		TEST_ASSERT_EQUAL_INT('a', fputc('a', fd));
+		TEST_ASSERT_EQUAL_INT('b', fputc('b', fd));
+		rewind(fd);
+
+		TEST_ASSERT_EQUAL_INT(EOF, fgetc(fd));
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+		TEST_ASSERT_EQUAL_INT(EOF, fgetc(fd));
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+	}
+	fclose(fd);
+
+	/* Try to write to file open for reading */
+	fd = fopen(stdpath, "r");
+	{
+		TEST_ASSERT_EQUAL_INT(EOF, fputc('a', fd));
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+		TEST_ASSERT_EQUAL_INT(EOF, fputc('a', fd));
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+	}
+	fclose(fd);
+}
+
+
+TEST(getput, getsputs_basic)
+{
+	/* reading/writing from file */
+	fd = fopen(stdpath, "w+");
+	{
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fputs(stdpath, fd));
+		rewind(fd);
+		TEST_ASSERT_NOT_NULL(fgets(buf, 16, fd));
+	}
+	fclose(fd);
+
+	/* reading from file not opened for reading */
+	fd = fopen(stdpath, "w");
+	{
+		TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fputs(stdpath, fd));
+		TEST_ASSERT_NULL(fgets(buf, 16, fd));
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+	}
+	fclose(fd);
+
+	fd = fopen(stdpath, "w+");
+}
+
+
+TEST(getput, getsputs_readonly)
+{
+	/* 
+	Upon successful completion, fputc() shall return the value it has written. 
+	Otherwise, it shall return EOF, the error indicator for the stream shall be set 
+	and errno shall be set to indicate the error. 
+	
+	https://github.com/phoenix-rtos/phoenix-rtos-project/issues/260
+	*/
+	TEST_IGNORE();
+
+	fd = fopen(stdpath, "r");
+	{
+		TEST_ASSERT_EQUAL_INT(EOF, fputs(stdpath, fd)); /* <posix incompliance> returns 0, should EOF */
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+		TEST_ASSERT_NOT_NULL(fgets(buf, 16, fd));
+		TEST_ASSERT_EQUAL_CHAR_ARRAY(stdpath, buf, sizeof(stdpath));
+		TEST_ASSERT_NULL(fgets(buf, 16, fd));
+	}
+	fclose(fd);
+}
+
+
+TEST(getput, ungetc_basic)
+{
+	/* standard usage of ungetc */
+	fd = fopen(stdpath, "w");
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fputs(stdpath, fd));
+	fclose(fd);
+	fd = fopen(stdpath, "r");
+	{
+		c = fgetc(fd);
+		TEST_ASSERT_EQUAL_INT((int)c, ungetc(c, fd));
+		TEST_ASSERT_EQUAL_PTR(buf, fgets(buf, sizeof(stdpath), fd));
+		TEST_ASSERT_EQUAL_STRING(stdpath, buf);
+	}
+	fclose(fd);
+
+	/*	EOF pushback test
+		If the value of c equals that of the macro EOF, 
+		the operation shall fail and the input stream shall be left unchanged. 
+	*/
+	fd = fopen(stdpath, "r");
+	{
+		TEST_ASSERT_EQUAL_INT(EOF, ungetc(EOF, fd));
+		TEST_ASSERT_EQUAL_INT(stdpath[0], fgetc(fd));
+	}
+	fclose(fd);
+}
+
+/*
+test group for:
+getline,
+*/
+TEST_GROUP(line);
+
+TEST_SETUP(line)
+{
+	/* file preparation */
+	fd = fopen(stdpath, "w+");
+	{
+		fputs("line1\nlineline2\nline3\n\n", fd);
+	}
+	fclose(fd);
+}
+
+
+TEST_TEAR_DOWN(line)
+{ /* empty */
+}
+
+
+TEST(line, getline_basic)
```

Comment:
Magic numbers (6, 7, 10, 11)

Reply:
Went for `define` approach. Now lines are separate defines and their use is more abstract. Thanks for pointing it out.